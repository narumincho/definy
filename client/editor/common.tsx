/* eslint-disable complexity */
import * as React from "react";
import {
  AccountIdDataOperation,
  AccountIdSelection,
  AccountIdValue,
  accountIdOperation,
} from "./accountId";
import { ImageSelection, ImageValue, imageOperation } from "./image";
import {
  ListDataOperation,
  ListSelection,
  ListValue,
  listOperation,
} from "./list";
import {
  NumberDataOperation,
  NumberSelection,
  NumberValue,
  numberOperation,
} from "./number";
import {
  ProductDataOperation,
  ProductSelection,
  ProductValue,
  productOperation,
} from "./product";
import {
  ProjectIdDataOperation,
  ProjectIdSelection,
  ProjectIdValue,
  projectIdOperation,
} from "./projectId";
import { SumDataOperation, SumSelection, SumValue, sumOperation } from "./sum";
import {
  TextDataOperation,
  TextSelection,
  TextValue,
  textOperation,
} from "./text";
import { TimeSelection, TimeValue, timeOperation } from "./time";
import {
  TypePartIdDataOperation,
  TypePartIdSelection,
  TypePartIdValue,
  typePartIdOperation,
} from "./typePartId";
import type { ElementOperation } from "./ElementOperation";
import { maybeMap } from "../../common/util";

export type Selection =
  | {
      tag: "product";
      value: ProductSelection;
    }
  | {
      tag: "list";
      value: ListSelection;
    }
  | {
      tag: "text";
      textSelection: TextSelection;
    }
  | {
      tag: "number";
      numberSelection: NumberSelection;
    }
  | {
      tag: "sum";
      sumSelection: SumSelection;
    }
  | {
      tag: "image";
      imageSelection: ImageSelection;
    }
  | {
      tag: "account";
      accountSelection: AccountIdSelection;
    }
  | {
      tag: "time";
      timeSelection: TimeSelection;
    }
  | {
      tag: "project";
      projectSelection: ProjectIdSelection;
    }
  | {
      tag: "typePartId";
      typePartIdSelection: TypePartIdSelection;
    };

const selectionProduct = (value: ProductSelection): Selection => ({
  tag: "product",
  value,
});
const selectionList = (value: ListSelection): Selection => ({
  tag: "list",
  value,
});

export type Value =
  | {
      type: "text";
      value: TextValue;
    }
  | {
      type: "number";
      value: NumberValue;
    }
  | {
      type: "sum";
      value: SumValue;
    }
  | {
      type: "accountId";
      value: AccountIdValue;
    }
  | {
      type: "time";
      value: TimeValue;
    }
  | {
      type: "list";
      value: ListValue;
    }
  | {
      type: "product";
      value: ProductValue;
    }
  | {
      type: "image";
      value: ImageValue;
    }
  | {
      type: "projectId";
      value: ProjectIdValue;
    }
  | {
      type: "typePartId";
      value: TypePartIdValue;
    };

export const textValue = (value: TextValue): Value => ({ type: "text", value });
export const numberValue = (value: NumberValue): Value => ({
  type: "number",
  value,
});
export const sumValue = (value: SumValue): Value => ({ type: "sum", value });
export const accountIdValue = (value: AccountIdValue): Value => ({
  type: "accountId",
  value,
});
export const timeValue = (value: TimeValue): Value => ({ type: "time", value });
export const listValue = (value: ListValue): Value => ({ type: "list", value });
export const productValue = (value: ProductValue): Value => ({
  type: "product",
  value,
});
export const imageValue = (value: ImageValue): Value => ({
  type: "image",
  value,
});
export const projectIdValue = (value: ProjectIdValue): Value => ({
  type: "projectId",
  value,
});
export const typePartIdValue = (value: TypePartIdValue): Value => ({
  type: "typePartId",
  value,
});

export type CommonDataOperation =
  | {
      tag: "text";
      textDataOperation: TextDataOperation;
    }
  | {
      tag: "number";
      numberDataOperation: NumberDataOperation;
    }
  | {
      tag: "sum";
      sumDataOperation: SumDataOperation;
    }
  | {
      tag: "account";
      accountDataOperation: AccountIdDataOperation;
    }
  | {
      tag: "project";
      projectDataOperation: ProjectIdDataOperation;
    }
  | {
      tag: "list";
      listDataOperation: ListDataOperation;
    }
  | {
      tag: "product";
      productDataOperation: ProductDataOperation;
    }
  | {
      tag: "typePartId";
      typePartIdDataOperation: TypePartIdDataOperation;
    };

const moveUp = (selection: Selection, value: Value): Selection | undefined => {
  if (selection.tag === "list" && value.type === "list") {
    return maybeMap(
      listOperation.moveUp(selection.value, value.value),
      selectionList
    );
  }
  if (selection.tag === "product" && value.type === "product") {
    return maybeMap(
      productOperation.moveUp(selection.value, value.value),
      selectionProduct
    );
  }
};

const moveDown = (
  selection: Selection,
  value: Value
): Selection | undefined => {
  if (selection.tag === "list" && value.type === "list") {
    return maybeMap(
      listOperation.moveDown(selection.value, value.value),
      selectionList
    );
  }
  if (selection.tag === "product" && value.type === "product") {
    return maybeMap(
      productOperation.moveDown(selection.value, value.value),
      selectionProduct
    );
  }
};

const moveFirstChild = (
  selection: Selection | undefined,
  value: Value
): Selection | undefined => {
  if (value.type === "list") {
    return maybeMap(
      listOperation.moveFirstChild(
        selection !== undefined && selection.tag === "list"
          ? selection.value
          : undefined,
        value.value
      ),
      selectionList
    );
  }
  if (value.type === "product") {
    return maybeMap(
      productOperation.moveFirstChild(
        selection !== undefined && selection.tag === "product"
          ? selection.value
          : undefined,
        value.value
      ),
      selectionProduct
    );
  }
};

const moveParent: ElementOperation<
  Selection,
  Value,
  CommonDataOperation
>["moveParent"] = (selection, value) => {
  if (selection.tag === "list" && value.type === "list") {
    return maybeMap(
      listOperation.moveParent(selection.value, value.value),
      selectionList
    );
  }
  if (selection.tag === "product" && value.type === "product") {
    return maybeMap(
      productOperation.moveParent(selection.value, value.value),
      selectionProduct
    );
  }
};

const CommonElementSelectionView: ElementOperation<
  Selection,
  Value,
  CommonDataOperation
>["selectionView"] = (props) => {
  if (props.value.type === "number") {
    return (
      <numberOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "number"
            ? props.selection.numberSelection
            : undefined
        }
        onRequestDataOperation={(numberDataOperation) =>
          props.onRequestDataOperation({ tag: "number", numberDataOperation })
        }
      />
    );
  }
  if (props.value.type === "text") {
    return (
      <textOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "text"
            ? props.selection.textSelection
            : undefined
        }
        onRequestDataOperation={(textDataOperation) =>
          props.onRequestDataOperation({ tag: "text", textDataOperation })
        }
      />
    );
  }
  if (props.value.type === "sum") {
    return (
      <sumOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "sum"
            ? props.selection.sumSelection
            : undefined
        }
        onRequestDataOperation={(sumDataOperation) =>
          props.onRequestDataOperation({
            tag: "sum",
            sumDataOperation,
          })
        }
      />
    );
  }
  if (props.value.type === "image") {
    return (
      <imageOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "image"
            ? props.selection.imageSelection
            : undefined
        }
        onRequestDataOperation={props.onRequestDataOperation}
      />
    );
  }
  if (props.value.type === "accountId") {
    return (
      <accountIdOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "account"
            ? props.selection.accountSelection
            : undefined
        }
        onRequestDataOperation={(accountDataOperation) => {
          props.onRequestDataOperation({
            tag: "account",
            accountDataOperation,
          });
        }}
      />
    );
  }
  if (props.value.type === "time") {
    return (
      <timeOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "time"
            ? props.selection.timeSelection
            : undefined
        }
        onRequestDataOperation={props.onRequestDataOperation}
      />
    );
  }
  if (props.value.type === "projectId") {
    return (
      <projectIdOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "project"
            ? props.selection.projectSelection
            : undefined
        }
        onRequestDataOperation={(projectDataOperation) =>
          props.onRequestDataOperation({ tag: "project", projectDataOperation })
        }
      />
    );
  }
  if (props.value.type === "list") {
    return (
      <listOperation.selectionView
        value={props.value.value}
        onChangeSelection={(listSelection) =>
          props.onChangeSelection(selectionList(listSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "list"
            ? props.selection.value
            : undefined
        }
        onRequestDataOperation={(listDataOperation) =>
          props.onRequestDataOperation({ tag: "list", listDataOperation })
        }
      />
    );
  }
  if (props.value.type === "product") {
    return (
      <productOperation.selectionView
        value={props.value.value}
        onChangeSelection={(productSelection) =>
          props.onChangeSelection(selectionProduct(productSelection))
        }
        selection={
          props.selection !== undefined && props.selection.tag === "product"
            ? props.selection.value
            : undefined
        }
        onRequestDataOperation={(productDataOperation) =>
          props.onRequestDataOperation({
            tag: "product",
            productDataOperation,
          })
        }
      />
    );
  }
  if (props.value.type === "typePartId") {
    return (
      <typePartIdOperation.selectionView
        value={props.value.value}
        onChangeSelection={(typePartIdSelection) =>
          props.onChangeSelection(typePartIdSelection)
        }
        selection={
          props.selection !== undefined && props.selection.tag === "typePartId"
            ? props.selection.typePartIdSelection
            : undefined
        }
        onRequestDataOperation={props.onRequestDataOperation}
      />
    );
  }
  return (
    <div>
      common で 値のパターンを記述忘れている
      {JSON.stringify(props.value)}
    </div>
  );
};

const CommonElementDetailView: ElementOperation<
  Selection,
  Value,
  CommonDataOperation
>["detailView"] = (props) => {
  if (props.value.type === "number") {
    return (
      <numberOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "number"
            ? props.selection.numberSelection
            : undefined
        }
        onRequestDataOperation={(numberDataOperation) =>
          props.onRequestDataOperation({ tag: "number", numberDataOperation })
        }
      />
    );
  }
  if (props.value.type === "text") {
    return (
      <textOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "text"
            ? props.selection.textSelection
            : undefined
        }
        onRequestDataOperation={(textDataOperation) =>
          props.onRequestDataOperation({
            tag: "text",
            textDataOperation,
          })
        }
      />
    );
  }
  if (props.value.type === "sum") {
    return (
      <sumOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "sum"
            ? props.selection.sumSelection
            : undefined
        }
        onRequestDataOperation={(sumDataOperation) =>
          props.onRequestDataOperation({
            tag: "sum",
            sumDataOperation,
          })
        }
      />
    );
  }
  if (props.value.type === "image") {
    return (
      <imageOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "image"
            ? props.selection.imageSelection
            : undefined
        }
        onRequestDataOperation={props.onRequestDataOperation}
      />
    );
  }
  if (props.value.type === "accountId") {
    return (
      <accountIdOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "account"
            ? props.selection.accountSelection
            : undefined
        }
        onRequestDataOperation={(accountDataOperation) =>
          props.onRequestDataOperation({ tag: "account", accountDataOperation })
        }
      />
    );
  }
  if (props.value.type === "projectId") {
    return (
      <projectIdOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "project"
            ? props.selection.projectSelection
            : undefined
        }
        onRequestDataOperation={(projectDataOperation) =>
          props.onRequestDataOperation({ tag: "project", projectDataOperation })
        }
      />
    );
  }
  if (props.value.type === "time") {
    return (
      <timeOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "time"
            ? props.selection.timeSelection
            : undefined
        }
        onRequestDataOperation={props.onRequestDataOperation}
      />
    );
  }
  if (props.value.type === "list") {
    return (
      <listOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "list"
            ? props.selection.value
            : undefined
        }
        onRequestDataOperation={(listDataOperation) =>
          props.onRequestDataOperation({
            tag: "list",
            listDataOperation,
          })
        }
      />
    );
  }
  if (props.value.type === "product") {
    return (
      <productOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "product"
            ? props.selection.value
            : undefined
        }
        onRequestDataOperation={(productDataOperation) =>
          props.onRequestDataOperation({
            tag: "product",
            productDataOperation,
          })
        }
      />
    );
  }
  if (props.value.type === "typePartId") {
    return (
      <typePartIdOperation.detailView
        value={props.value.value}
        selection={
          props.selection !== undefined && props.selection.tag === "typePartId"
            ? props.selection.typePartIdSelection
            : undefined
        }
        onRequestDataOperation={props.onRequestDataOperation}
      />
    );
  }
  return <div>選択時しているものの構造が壊れている</div>;
};

export const commonElement: ElementOperation<
  Selection,
  Value,
  CommonDataOperation
> = {
  moveUp,
  moveDown,
  moveFirstChild,
  moveParent,
  selectionView: CommonElementSelectionView,
  detailView: CommonElementDetailView,
};
