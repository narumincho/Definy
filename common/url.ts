import * as d from "../data";
import { nowMode } from "../out";

export const origin =
  nowMode === d.Mode.Develop ? `http://localhost:2520` : "https://definy.app";

const languageQueryKey = "hl";
export const defaultLanguage: d.Language = "English";

export const urlDataAndAccountTokenToUrl = (
  urlData: d.UrlData,
  accountToken: d.Maybe<d.AccountToken>
): URL => {
  const url = new URL(origin);
  url.pathname = locationToPath(urlData.location);
  url.searchParams.append(
    languageQueryKey,
    languageToIdString(urlData.language)
  );
  if (accountToken._ === "Just") {
    url.hash = "account-token=" + (accountToken.value as string);
  }
  return url;
};

const locationToPath = (location: d.Location): string => {
  switch (location._) {
    case "Home":
      return "/";
    case "CreateProject":
      return "/create-project";
    case "Account":
      return "/user/" + (location.accountId as string);
    case "Project":
      return "/project/" + (location.projectId as string);
    case "Setting":
      return "/setting/";
    case "About":
      return "/about";
    case "Debug":
      return "/debug";
    case "TypePart":
      return "/type-part/" + location.typePartId;
  }
};

const languageToIdString = (language: d.Language): string => {
  switch (language) {
    case "English":
      return "en";
    case "Japanese":
      return "ja";
    case "Esperanto":
      return "eo";
  }
};

/**
 * URLのパスを場所のデータに変換する
 * @param url `https://definy.app/project/580d8d6a54cf43e4452a0bba6694a4ed?hl=ja` のようなURL
 */
export const urlDataAndAccountTokenFromUrl = (
  url: URL
): { urlData: d.UrlData; accountToken: d.Maybe<d.AccountToken> } => {
  const languageId = url.searchParams.get(languageQueryKey);
  const language: d.Language =
    languageId === null ? defaultLanguage : languageFromIdString(languageId);
  return {
    urlData: {
      location: locationFromUrl(url.pathname),
      language,
    },
    accountToken: accountTokenFromUrl(url.hash),
  };
};

const locationFromUrl = (pathName: string): d.Location => {
  if (pathName === "/create-project") {
    return d.Location.CreateProject;
  }
  if (pathName === "/about") {
    return d.Location.About;
  }
  if (pathName === "/debug") {
    return d.Location.Debug;
  }
  if (pathName === "/setting") {
    return d.Location.Setting;
  }
  const projectResult = pathName.match(/^\/project\/(?<id>[0-9a-f]{32})$/u);
  if (projectResult !== null && projectResult.groups !== undefined) {
    return d.Location.Project(projectResult.groups.id as d.ProjectId);
  }
  const userResult = pathName.match(/^\/user\/(?<id>[0-9a-f]{32})$/u);
  if (userResult !== null && userResult.groups !== undefined) {
    return d.Location.Account(userResult.groups.id as d.AccountId);
  }
  const typePartResult = pathName.match(/^\/type-part\/(?<id>[0-9a-f]{32})$/u);
  if (typePartResult !== null && typePartResult.groups !== undefined) {
    return d.Location.TypePart(typePartResult.groups.id as d.TypePartId);
  }
  return d.Location.Home;
};

const languageFromIdString = (languageAsString: string): d.Language => {
  switch (languageAsString) {
    case "ja":
      return "Japanese";
    case "en":
      return "English";
    case "eo":
      return "Esperanto";
  }
  return defaultLanguage;
};

const accountTokenFromUrl = (hash: string): d.Maybe<d.AccountToken> => {
  const matchResult = hash.match(/account-token=(?<token>[0-9a-f]{64})/u);
  if (matchResult === null || matchResult.groups === undefined) {
    return d.Maybe.Nothing();
  }
  return d.Maybe.Just(matchResult.groups.token as d.AccountToken);
};

export const iconUrl: URL = new URL(`${origin}/icon`);

export const pngFileUrl = (fileHash: d.ImageHash): URL =>
  new URL(`${origin}/pngFile/${fileHash}`);

export const apiUrl = (apiName: string): URL =>
  new URL(`${origin}/api/${apiName}`);

export const logInRedirectUri = (
  openIdConnectProvider: d.OpenIdConnectProvider
): string => `${origin}/logInCallback/${openIdConnectProvider}`;
