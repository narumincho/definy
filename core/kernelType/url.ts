import * as c from "./codec";
import * as identifer from "../../gen/jsTs/identifer";
import * as string from "./string";
import * as ts from "../../gen/jsTs/data";
import * as tsUtil from "../../gen/jsTs/util";
import * as util from "../util";

const name = identifer.fromString("Url");

export const type = ts.Type.ScopeInGlobal(name);

export const encodeDefinitionStatementList = (
  valueVar: ts.Expr
): ReadonlyArray<ts.Statement> => {
  return [
    ts.Statement.Return(
      util.callEncode(
        string.codec(),
        tsUtil.callMethod(valueVar, "toString", [])
      )
    ),
  ];
};

export const decodeDefinitionStatementList = (
  parameterIndex: ts.Expr,
  parameterBinary: ts.Expr
): ReadonlyArray<ts.Statement> => {
  const resultAndNextIndexAsStringName = identifer.fromString(
    "resultAndNextIndexAsString"
  );
  const resultAndNextIndexAsStringVar = ts.Expr.Variable(
    resultAndNextIndexAsStringName
  );
  return [
    ts.Statement.VariableDefinition({
      isConst: true,
      name: resultAndNextIndexAsStringName,
      type: c.decodeReturnType(string.type),
      expr: util.callDecode(string.codec(), parameterIndex, parameterBinary),
    }),
    c.returnStatement(
      ts.Expr.New({
        expr: ts.Expr.GlobalObjects(identifer.fromString("URL")),
        parameterList: [c.getResult(resultAndNextIndexAsStringVar)],
      }),
      c.getNextIndex(resultAndNextIndexAsStringVar)
    ),
  ];
};
