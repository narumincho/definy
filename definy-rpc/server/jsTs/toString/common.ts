import { TsIdentifier } from "../identifier.ts";

export const documentToString = (document: string): string => {
  const documentTrimmed = document.trim();
  return documentTrimmed === ""
    ? ""
    : "\n/**\n" +
        documentTrimmed
          .split("\n")
          .map((line) =>
            line === "" ? " *" : " * " + line.replace(/\*\//gu, "* /")
          )
          .join("\n") +
        "\n */\n";
};

/**
 * 文字列を`"`で囲んでエスケープする
 */
export const stringLiteralValueToString = (value: string): string =>
  '"' +
  value
    .replace(/\\/gu, "\\\\")
    .replace(/"/gu, '\\"')
    .replace(/\r\n|\n/gu, "\\n") +
  '"';

/**
 * 型パラメーターを文字列にする `<T extends unknown>` `<ok extends unknown, error extends unknown>`
 * extends unknown をつけた理由はJSXでも解釈できるようにするため
 */
export const typeParameterListToString = (
  typeParameterList: ReadonlyArray<TsIdentifier>
): string => {
  if (typeParameterList.length === 0) {
    return "";
  }
  return (
    "<" +
    typeParameterList
      .map((typeParameter) => typeParameter + " extends unknown")
      .join(", ") +
    ">"
  );
};

export const indentNumberToString = (indent: number): string =>
  "  ".repeat(indent);
