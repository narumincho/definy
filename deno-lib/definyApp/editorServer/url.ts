export const toOgpImageUrl = (text: string): string => {
  const params = new URLSearchParams();
  params.append("text", text);
  return "/clock-ogp?" + params.toString();
};

export const extractOgpUrl = (url: URL): string | undefined => {
  const text = url.searchParams.get("text");
  if (url.pathname === "/clock-ogp" && typeof text === "string") {
    return text;
  }
  return undefined;
};
