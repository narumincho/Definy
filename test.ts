import * as commonUrl from "./common/url";
import * as data from "./data";
import { strict as strictAssert } from "assert";

commonUrl.urlDataAndAccountTokenFromUrl(new URL("https://definy.app/")).urlData,
  {
    clientMode: "Release",
    location: data.Location.Home,
    language: "English",
  },
  "https://definy.app/ is Home in English";

strictAssert.deepEqual(
  commonUrl.urlDataAndAccountTokenFromUrl(
    new URL("https://definy.app/project/580d8d6a54cf43e4452a0bba6694a4ed?hl=ja")
  ).urlData,
  {
    clientMode: "Release",
    location: data.Location.Project(
      "580d8d6a54cf43e4452a0bba6694a4ed" as data.ProjectId
    ),
    language: "Japanese",
  }
);
{
  const url = new URL(
    "http://localhost:2520/user/580d8d6a54cf43e4452a0bba6694a4ed?hl=eo#account-token=f81919b78537257302b50f776b77a90b984cc3d75fa899f9f460ff972dcc8cb0"
  );
  strictAssert.deepEqual(
    commonUrl.urlDataAndAccountTokenFromUrl(url).urlData,
    {
      clientMode: "DebugMode",
      location: data.Location.Account(
        "580d8d6a54cf43e4452a0bba6694a4ed" as data.AccountId
      ),
      language: "Esperanto",
    },
    "local host"
  );
}
{
  const url = new URL(
    "http://localhost:2520/user/580d8d6a54cf43e4452a0bba6694a4ed?hl=eo#account-token=f81919b78537257302b50f776b77a90b984cc3d75fa899f9f460ff972dcc8cb0"
  );
  strictAssert.deepEqual(
    commonUrl.urlDataAndAccountTokenFromUrl(url).accountToken,
    data.Maybe.Just(
      "f81919b78537257302b50f776b77a90b984cc3d75fa899f9f460ff972dcc8cb0" as data.AccountToken
    ),
    "accountToken"
  );
}
{
  const languageAndLocation: data.UrlData = {
    location: data.Location.Account(
      "580d8d6a54cf43e4452a0bba6694a4ed" as data.AccountId
    ),
    language: "Esperanto",
  };
  const url = commonUrl.urlDataAndAccountTokenToUrl(
    languageAndLocation,
    data.Maybe.Nothing()
  );
  const decodedLanguageAndLocation: data.UrlData = commonUrl.urlDataAndAccountTokenFromUrl(
    url
  ).urlData;
  strictAssert.deepEqual(
    languageAndLocation,
    decodedLanguageAndLocation,
    "encode, decode user url"
  );
}
