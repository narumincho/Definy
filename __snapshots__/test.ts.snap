// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test add attribute 1`] = `
Object {
  "elementUpdateDiff": Object {
    "children": Object {
      "kind": "skip",
    },
    "class": undefined,
    "id": "newId",
    "tag": "div",
  },
  "key": "_",
  "kind": "update",
}
`;

exports[`test add element 1`] = `
Object {
  "children": Array [
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "a",
      "kind": "update",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "b",
      "kind": "update",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "newC",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "c",
      "kind": "insert",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test change key order 1`] = `
Object {
  "children": Array [
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "b",
      "kind": "update",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "c",
      "kind": "update",
    },
    Object {
      "key": "a",
      "kind": "replace",
      "newElement": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "A",
        },
        "class": "css-0",
        "click": null,
        "id": "A",
        "tag": "div",
      },
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test complex diff 1`] = `
Object {
  "children": Array [
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "setText",
          "text": "A",
        },
        "class": undefined,
        "id": "",
        "tag": "div",
      },
      "key": "A",
      "kind": "update",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "B",
      "kind": "update",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "K",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "K",
      "kind": "insert",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "children": Array [
            Object {
              "elementUpdateDiff": Object {
                "children": Object {
                  "kind": "skip",
                },
                "class": undefined,
                "id": undefined,
                "tag": "div",
              },
              "key": "text",
              "kind": "update",
            },
            Object {
              "kind": "delete",
            },
          ],
          "kind": "childDiffList",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "C",
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "S",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "S",
      "kind": "insert",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "E",
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test delete all children 1`] = `
Object {
  "children": Array [
    Object {
      "kind": "delete",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "X",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "x",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "Y",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "y",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "Z",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "z",
      "kind": "insert",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test delete attribute 1`] = `
Object {
  "elementUpdateDiff": Object {
    "children": Object {
      "kind": "skip",
    },
    "class": undefined,
    "id": "",
    "tag": "div",
  },
  "key": "_",
  "kind": "update",
}
`;

exports[`test delete element 1`] = `
Object {
  "children": Array [
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "a",
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "c",
      "kind": "update",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test different text return setText 1`] = `
Object {
  "kind": "setText",
  "text": "B",
}
`;

exports[`test insert element 1`] = `
Object {
  "children": Array [
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "a",
      "kind": "update",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "new",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "new",
      "kind": "insert",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "b",
      "kind": "update",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test replace all 1`] = `
Object {
  "children": Array [
    Object {
      "kind": "delete",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "X",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "x",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "Y",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "y",
      "kind": "insert",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "Z",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "z",
      "kind": "insert",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test replace element 1`] = `
Object {
  "children": Array [
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "a",
      "kind": "update",
    },
    Object {
      "kind": "delete",
    },
    Object {
      "element": Object {
        "children": Object {
          "tag": Symbol(Children - Text),
          "value": "new",
        },
        "class": "css-0",
        "click": null,
        "id": "",
        "tag": "div",
      },
      "key": "new",
      "kind": "insert",
    },
    Object {
      "elementUpdateDiff": Object {
        "children": Object {
          "kind": "skip",
        },
        "class": undefined,
        "id": undefined,
        "tag": "div",
      },
      "key": "c",
      "kind": "update",
    },
  ],
  "kind": "childDiffList",
}
`;

exports[`test same text return skip 1`] = `
Object {
  "kind": "skip",
}
`;
